stable_baselines3/her/her.py

```diff
@@ -0,0 +1,416 @@
+                # do as many gradients steps as steps performed during the rollout
+                gradient_steps = self.gradient_steps if self.gradient_steps > 0 else rollout.episode_timesteps
+                self.train(batch_size=self.batch_size, gradient_steps=gradient_steps)
+
+        callback.on_training_end()
+
+        return self
+
+    def collect_rollouts(
+        self,
+        env: VecEnv,
+        callback: BaseCallback,
+        n_episodes: int = 1,
+        n_steps: int = -1,
+        action_noise: Optional[ActionNoise] = None,
+        learning_starts: int = 0,
+        replay_buffer: Union[ReplayBuffer, HerReplayBuffer] = None,
+        log_interval: Optional[int] = None,
+    ) -> RolloutReturn:
+        """
+        Collect experiences and store them into a ReplayBuffer.
+
+        :param env: (VecEnv) The training environment
+        :param callback: (BaseCallback) Callback that will be called at each step
+            (and at the beginning and end of the rollout)
+        :param n_episodes: (int) Number of episodes to use to collect rollout data
+            You can also specify a ``n_steps`` instead
+        :param n_steps: (int) Number of steps to use to collect rollout data
+            You can also specify a ``n_episodes`` instead.
+        :param action_noise: (Optional[ActionNoise]) Action noise that will be used for exploration
+            Required for deterministic policy (e.g. TD3). This can also be used
+            in addition to the stochastic policy for SAC.
+        :param learning_starts: (int) Number of steps before learning for the warm-up phase.
+        :param replay_buffer: (ReplayBuffer or HerReplayBuffer)
+        :param log_interval: (int) Log data every ``log_interval`` episodes
+        :return: (RolloutReturn)
+        """
+        episode_rewards, total_timesteps = [], []
+        total_steps, total_episodes = 0, 0
+
+        assert isinstance(env, VecEnv), "You must pass a VecEnv"
+        assert env.num_envs == 1, "OffPolicyAlgorithm only support single environment"
+
+        if self.use_sde:
+            self.model.actor.reset_noise()
+
+        callback.on_rollout_start()
+        continue_training = True
+
+        while total_steps < n_steps or total_episodes < n_episodes:
+            done = False
+            episode_reward, episode_timesteps = 0.0, 0
+
+            while not done:
+                # concatenate observation and (desired) goal
+                observation = self.model._last_obs
+                self.model._last_obs = np.concatenate([observation["observation"], observation["desired_goal"]], axis=1)
+
+                if self.use_sde and self.sde_sample_freq > 0 and total_steps % self.sde_sample_freq == 0:
+                    # Sample a new noise matrix
+                    self.model.actor.reset_noise()
+
+                # Select action randomly or according to policy
+                action, buffer_action = self.model._sample_action(learning_starts, action_noise)
```

Comment:
after thinking more about it, I think we need to define `__get__attr_` to automaticaly retrieve the attribute from `self.model` if present. This would allow to write directly `self._sample_action()` .